# Modeling Scratchwork

## Loading in Libraries and Data

In [1]:
# Packages for data cleaning, plotting, and manipulation

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

# scikit-learn libraries/functions/classes

from sklearn.tree import DecisionTreeClassifier
from sklearn.preprocessing import OneHotEncoder, LabelEncoder, StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score
from sklearn.metrics import accuracy_score, precision_score, recall_score, \
                            ConfusionMatrixDisplay, classification_report, \
                            confusion_matrix
from sklearn.linear_model import LogisticRegression, RidgeCV
from sklearn.impute import SimpleImputer, KNNImputer
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier, \
                             StackingClassifier, ExtraTreesClassifier
from sklearn.dummy import DummyClassifier
from sklearn.compose import ColumnTransformer
from xgboost import XGBClassifier

In [2]:
# Shows *all* columns in dataframe, i.e. does not truncate horizontally;
# feel free to comment out if undesired

pd.set_option('display.max_columns', None)

In [3]:
# Importing training data
train_val = pd.read_csv('../data/training_set_values.csv')

# Only using `status_group` column from label set, to
# avoid duplicating `id` column
train_label = pd.read_csv('../data/training_set_labels.csv',
                             usecols = ['status_group'])

In [4]:
# Concatenating separate .csv files
df = pd.concat(objs = [train_val, train_label],
               axis = 1)

## Preprocessing

### Feature Selection

In [5]:
# Dropping columns determined to be either irrelevant or
# superfluous in exploratory analysis

cols_to_drop = [
    'id',  # unique identifier, not useful for modeling
    'date_recorded',  # superfluous information, too many unique records
    'recorded_by',  # no unique information + no unique values across 59.4k rows
    'funder',   
    'installer',  # large number of unique values (see also `funder`);
                  # may be added back in later
    'wpt_name',  # identifier column, not useful for modeling
    'num_private',  # data dict. does not provide details for this column
    'subvillage',  # too many unique values - uninformative for modeling
    'region_code',  # redundant information vis-a-visa the simpler `region`
    'district_code',  # may be added back in later
    'lga',
    'ward',  # redundant location data (with `lga`)
    'scheme_management',  # may be added back in later
    'scheme_name',  # large number of nulls, redundant vis-a-vis `scheme_management`
    'extraction_type',
    'extraction_type_group',  # using `extraction_type_class` for generalized info
    'management',
    'management_group',  # may be added back in later
    'payment',  # identical information to `payment_type`
    'water_quality',  # comparable information to `quality_group` - redundant
    'quantity_group',  # identical information to `quantity` - redundant
    'source',  # redundant with other `source_` columns
    'waterpoint_type'  # used `waterpoint_type_group` instead
]

In [6]:
df = df.drop(columns = cols_to_drop).copy()

### Pipeline Creation

Necessary modifications for modeling, to be written into pipelines:

- One-hot encoding:
    - `basin`
    - `extraction_type_class`
    - `payment_type`
    - `permit`
    - `public_meeting`
    - `quality_group`
    - `quantity`
    - `region`
    - `source_type`
    - `source_class`
    - `waterpoint_type_group`
- Numerical scaling:
    - `population` - (impute zeroes with median?)
    - `gps_height` - impute zeroes with median
    - `latitude` / `longitude` - impute zeroes with mean
    - `construction_year` - use KNN imputing

In [7]:
# Subpipes for imputing median values - to be used for `latitude` and `longitude`
subpipe_lat      = Pipeline(steps=[('num_impute', SimpleImputer(missing_values = -2.000000e-08,
                                                                strategy = 'median')),
                                   ('ss', StandardScaler())])

subpipe_long     = Pipeline(steps=[('num_impute', SimpleImputer(missing_values = 0.000000,
                                                                strategy = 'median')),
                                   ('ss', StandardScaler())])


# Subpipe for imputing median values
subpipe_num      = Pipeline(steps=[('num_impute', SimpleImputer(strategy = 'median')),
                                   ('ss', StandardScaler())])

# Subpipe for `construction_year`
subpipe_year     = Pipeline(steps=[('num_impute', SimpleImputer(missing_values = 0,
                                                                strategy = 'median')),
                                   ('ss', StandardScaler())])

# Subpipe for categorical features including `basin`, `payment_type`
subpipe_cat      = Pipeline(steps=[('freq_imputer_nan', SimpleImputer(strategy = 'most_frequent')),
                                   ('freq_imputer_unk', SimpleImputer(strategy = 'most_frequent',
                                                                      missing_values = 'unknown')),
                                   ('ohe', OneHotEncoder(drop = 'if_binary',
                                                         sparse = False,
                                                         handle_unknown = 'ignore'))])


# Subpipe for the target column, `status_group`
subpipe_label    = Pipeline(steps=[('le', LabelEncoder())])

In [8]:
# Columns to be passed through numerical pipeline
num_cols = ['amount_tsh',
            'gps_height',
            'population']

# Columns to be passed through categorical pipeline
cat_cols = ['basin',
            'region',
            'payment_type',
            'quantity',
            'quality_group',
            'permit',
            'public_meeting',
            'extraction_type_class',
            'source_type',
            'source_class',
            'waterpoint_type_group']

In [9]:
ct = ColumnTransformer(transformers = [
    ('subpipe_num', subpipe_num, num_cols),
    ('subpipe_year', subpipe_year, ['construction_year']),
    ('subpipe_long', subpipe_long, ['longitude']),
    ('subpipe_lat', subpipe_lat, ['latitude']),
    ('subpipe_cat', subpipe_cat, cat_cols)],
                       remainder = 'passthrough')

# ('subpipe_label', subpipe_label, ['status_group'])

### Train/Test Split and Initial Preparation for ML

In [10]:
# Splitting DataFrame into features/values DataFrame
# (i.e. `X`) and labels series (`y`)

X = df.drop('status_group', axis = 1)
y = df['status_group']

In [11]:
# Splitting internal training data into separate
# training and test sets for (eventual) internal validation

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.33, random_state = 138)

## Modeling

### `DummyClassifier`

In [12]:
dummy_model_pipe = Pipeline(steps=[
    ('ct', ct),
    ('dummy', DummyClassifier())
])

In [13]:
# Fit on training data
dummy_model_pipe.fit(X_train, y_train)

# Score on training data
dummy_model_pipe.score(X_train, y_train)

0.5420875420875421

scikit-learn's `DummyClassifier` predicts on the training data with an accuracy score of ~0.542, equal to the proportion of the **most frequent class** (`functional`). This is because, as a dummy model, it predicts `functional` (the most frequent value) every time.

We'll be looking to improve on that 54.2% accuracy in future models.

### Simple Model - `LogisticRegression`

#### Default hyperparameters

In [14]:
# Default arguments - max iterations set to 100
logreg_model_simple = Pipeline(steps=[
    ('ct', ct),
    ('logreg', LogisticRegression(random_state = 138))
])

In [15]:
logreg_model_simple.fit(X_train, y_train)

logreg_model_simple.score(X_train, y_train)

C:\Users\toast\anaconda3\envs\learn-env\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


0.7217950650786472

> First simple model is giving us an accuracy rate of **~72%** - and a `ConvergenceWarning`.

##### Simple Model 1.1 - increasing `max_iter`

In [16]:
# Increasing `max_iter` to 1000
logreg_model_more_iter = Pipeline(steps=[
    ('ct', ct),
    ('logreg', LogisticRegression(random_state = 138,
                                  max_iter = 1000))
])

In [17]:
logreg_model_more_iter.fit(X_train, y_train)

logreg_model_more_iter.score(X_train, y_train)

0.7214935423890648

> Increasing `max_iter` from 100 to 1,000 solved the `ConvergenceWarning`; it also resulted in a *slightly* different accuracy score.

#### Implementing new solver (`saga`)

In [18]:
# Adjusting solver - changed from 'lbfgs' to 'saga'
# also dropping `max_iter` back down to default
logreg_model_saga = Pipeline(steps=[
    ('ct', ct),
    ('logreg', LogisticRegression(random_state = 138,
                                  solver = 'saga'))
])

In [19]:
logreg_model_saga.fit(X_train, y_train)

logreg_model_saga.score(X_train, y_train)

C:\Users\toast\anaconda3\envs\learn-env\lib\site-packages\sklearn\linear_model\_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


0.7205889743203177

> Ever-so-slightly worse accuracy score when the solver was changed from `lbfgs` to `saga`. But *still* getting that `ConvergenceWarning`.

#### Adding `elasticnet` penalty

In [20]:
# Changed penalty to 'elasticnet', set 'l1_ratio' to 0.6
# Reduced max_iter to 10 -  not much sacrifice in accuracy,
# better processing time w/ ensemble methods
logreg_model_saga_elasnet = Pipeline(steps=[
    ('ct', ct),
    ('logreg', LogisticRegression(penalty = 'elasticnet',
                                  l1_ratio = 0.6,
                                  solver = 'saga',
                                  random_state = 138,
                                  max_iter = 10))
])

In [21]:
logreg_model_saga_elasnet.fit(X_train, y_train)

logreg_model_saga_elasnet.score(X_train, y_train)

C:\Users\toast\anaconda3\envs\learn-env\lib\site-packages\sklearn\linear_model\_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


0.7186793306196291

These scores aren't moving very much with manual tuning.

#### Reducing regularization (`C`)

Let's try tuning *one* more hyperparameter in the `LogisticRegression` class before moving on to something else. We'll also drop `max_iter` back down to the default 100 to reduce the amount of processing time needed in future calculations.

In [22]:
# Reduced regularization by inflating C parameter
logreg_less_reg = Pipeline(steps=[
    ('ct', ct),
    ('logreg', LogisticRegression(C = 1e5,
                                  penalty = 'elasticnet',
                                  l1_ratio = 0.6,
                                  solver = 'saga',
                                  random_state = 138))
])

In [23]:
logreg_less_reg.fit(X_train, y_train)

logreg_less_reg.score(X_train, y_train)

C:\Users\toast\anaconda3\envs\learn-env\lib\site-packages\sklearn\linear_model\_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


0.7206392281019147

No better when the `C` value is set slightly higher, i.e. reduced regularization.

We're capping out at around **~72% accuracy** with various logistic regression models using this set of features. We're also still getting the `ConvergenceWarning`.

Let's look at a different algorithm.

### `DecisionTreeClassifier` - another simple model

#### Default hyperparameters

In [75]:
dtc_sample_pipe = Pipeline(steps=[
    ('ct', ct),
    ('dtc', DecisionTreeClassifier(random_state = 138))
])

In [76]:
dtc_sample_pipe.fit(X_train, y_train)

dtc_sample_pipe.score(X_train, y_train)

0.9942208151163375

#### Limiting `max_depth`

In [61]:
dtc_model_pipe = Pipeline(steps=[
    ('ct', ct),
    ('dtc', DecisionTreeClassifier(max_depth = 50,
                                   random_state = 138))
])

In [62]:
dtc_model_pipe.fit(X_train, y_train)

dtc_model_pipe.score(X_train, y_train)

0.9942208151163375

#### Further reducing `max_depth`

In [63]:
dtc_pipe_two = Pipeline(steps=[
    ('ct', ct),
    ('dtc', DecisionTreeClassifier(max_depth = 25,
                                   random_state = 138))
])

In [64]:
dtc_pipe_two.fit(X_train, y_train)

dtc_pipe_two.score(X_train, y_train)

0.9664807276747576

#### ...even further...

In [65]:
dtc_pipe_three = Pipeline(steps=[
    ('ct', ct),
    ('dtc', DecisionTreeClassifier(max_depth = 15,
                                   random_state = 138))
])

In [66]:
dtc_pipe_three.fit(X_train, y_train)

dtc_pipe_three.score(X_train, y_train)

0.8317503392130258

We're getting better numbers with `DecisionTreeClassifier` than with `LogisticRegression`, but we might *also* be overfitting with a decision tree. Let's play around with some other tree-based methods.

### `RandomForestClassifier`

#### Starting small: `max_depth = 10`

In [67]:
# Simple RFC - minimal changes from default hyperparams
# Starting small w/ max_depth = 10
rfc_model_pipe = Pipeline(steps=[
    ('ct', ct),
    ('rfc', RandomForestClassifier(max_depth = 10,
                                   random_state = 138))
])

In [68]:
rfc_model_pipe.fit(X_train, y_train)

rfc_model_pipe.score(X_train, y_train)

0.7584300718629077

> Accuracy score on the RFC model with default parameters is around **76%**, already nearly a 3.5% increase from our best logistic regression model.

#### `max_features = sqrt`

In [69]:
rfc_pipe_two = Pipeline(steps=[
    ('ct', ct),
    ('rfc', RandomForestClassifier(max_features = 'sqrt',
                                   max_depth = 10,
                                   random_state = 138))
])

In [70]:
rfc_pipe_two.fit(X_train, y_train)

rfc_pipe_two.score(X_train, y_train)

0.7584300718629077

> Very similar accuracy score (76%) to our first RFC model, even after modifying `max_features` hyperparameter.

#### Slight increase to `max_depth`

In [71]:
rfc_pipe_three = Pipeline(steps=[
    ('ct', ct),
    ('rfc', RandomForestClassifier(max_features = 'sqrt',
                                   max_depth = 15,
                                   random_state = 138))
])

In [72]:
rfc_pipe_three.fit(X_train, y_train)

rfc_pipe_three.score(X_train, y_train)

0.8390120106538017

In [73]:
print(classification_report(y_true = y_train,
                            y_pred = rfc_pipe_three.predict(X_train)))

                         precision    recall  f1-score   support

             functional       0.79      0.97      0.87     21574
functional needs repair       0.92      0.24      0.39      2921
         non functional       0.93      0.77      0.84     15303

               accuracy                           0.84     39798
              macro avg       0.88      0.66      0.70     39798
           weighted avg       0.85      0.84      0.82     39798



In [74]:
confusion_matrix(y_true = y_train,
                 y_pred = rfc_pipe_three.predict(X_train))

array([[20948,    36,   590],
       [ 1914,   715,   292],
       [ 3546,    29, 11728]], dtype=int64)

Increasing `max_depth` by 50% resulted in an ~8% increase in our accuracy score. It's possible that we're overfitting here, but an ensembled method with other algorithms might be able to blunt that overfitting.

### `StackingClassifier` experimentation

#### Choosing estimators

In [32]:
estimators = [
    ('logreg_model', logreg_model_saga_elasnet),
    ('rfc_model', rfc_model_pipe)
]

sc_model_pipe = StackingClassifier(estimators)

In [33]:
sc_model_pipe.fit(X_train, y_train)

sc_model_pipe.score(X_train, y_train)

C:\Users\toast\anaconda3\envs\learn-env\lib\site-packages\sklearn\linear_model\_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
C:\Users\toast\anaconda3\envs\learn-env\lib\site-packages\sklearn\linear_model\_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
C:\Users\toast\anaconda3\envs\learn-env\lib\site-packages\sklearn\linear_model\_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
C:\Users\toast\anaconda3\envs\learn-env\lib\site-packages\sklearn\linear_model\_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
C:\Users\toast\anaconda3\envs\learn-env\lib\site-packages\sklearn\linear_model\_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
C:\Users\toast\anaconda3\envs\learn

0.78373285089703

Our third logistic regression model - `logreg_model_saga_elasnet` - and our default RFC model - `rfc_model_pipe` - yielded an accuracy rate of **~78%** when stacked.

#### Cross-validating the stacked classifier

In [34]:
cross_val_score(sc_model_pipe, X_train, y_train)

C:\Users\toast\anaconda3\envs\learn-env\lib\site-packages\sklearn\linear_model\_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
C:\Users\toast\anaconda3\envs\learn-env\lib\site-packages\sklearn\linear_model\_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
C:\Users\toast\anaconda3\envs\learn-env\lib\site-packages\sklearn\linear_model\_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
C:\Users\toast\anaconda3\envs\learn-env\lib\site-packages\sklearn\linear_model\_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
C:\Users\toast\anaconda3\envs\learn-env\lib\site-packages\sklearn\linear_model\_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
C:\Users\toast\anaconda3\envs\learn

array([0.76595477, 0.76570352, 0.76281407, 0.75574821, 0.76504586])

Cross-validation scores in the `0.76` range, and consistent scores across the fold.

### `ExtraTreesClassifier`

#### Default hyperparameters

In [35]:
# Default hyperparameters
extra_trees_pipe = Pipeline(steps=[
    ('ct', ct),
    ('etc', ExtraTreesClassifier(random_state=138))
])

In [36]:
extra_trees_pipe.fit(X_train, y_train)

extra_trees_pipe.score(X_train, y_train)

0.9942208151163375

Likely *way* overfit. Let's tune those hyperparameters.

#### Gridsearching the `ExtraTreesClassifier`

In [37]:
# New ETC classifier for gridsearching
etc_pipe_gs = Pipeline(steps=[
    ('ct', ct),
    ('etc', ExtraTreesClassifier(random_state=138))
])

In [38]:
parameters = {'etc__criterion': ['gini', 'entropy'],
              'etc__min_samples_leaf': [1, 5, 10],
              'etc__max_depth': [5, 10, 20, 30]}

gs_etc = GridSearchCV(estimator=etc_pipe_gs,
                      param_grid=parameters,
                      cv=5)

In [39]:
gs_etc.fit(X_train, y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('ct',
                                        ColumnTransformer(remainder='passthrough',
                                                          transformers=[('subpipe_num',
                                                                         Pipeline(steps=[('num_impute',
                                                                                          SimpleImputer(strategy='median')),
                                                                                         ('ss',
                                                                                          StandardScaler())]),
                                                                         ['amount_tsh',
                                                                          'gps_height',
                                                                          'population']),
                                                              

Well, no need to run *that* cell ever again.

In [40]:
gs_etc.best_params_

{'etc__criterion': 'entropy', 'etc__max_depth': 20, 'etc__min_samples_leaf': 1}

In [41]:
gs_etc.best_score_

0.7874767827073551

In [42]:
# Modifying the pipeline and classifier to match
# best parameters found in gridsearch

etc_pipe_gs = Pipeline(steps=[
    ('ct', ct),
    ('etc', ExtraTreesClassifier(max_depth = 20,
                                 random_state = 138))
])

In [43]:
etc_pipe_gs.fit(X_train, y_train)

etc_pipe_gs.score(X_train, y_train)

0.900547766219408

Hm, it looks like we're probably still overfitting with the `ExtraTreesClassifier`, even after gridsearching.

### `GradientBoostingClassifier`

#### Default hyperparameters

In [44]:
gbc_model_pipe = Pipeline(steps=[
    ('ct', ct),
    ('gbc', GradientBoostingClassifier(random_state = 138))
])

In [45]:
gbc_model_pipe.fit(X_train, y_train)

gbc_model_pipe.score(X_train, y_train)

0.750942258404945

A decent score with the default hyperparameters. Let's try a more powerful gradient boosting tool.

### `XGBoost`

Let's give this a swing...

In [46]:
xgb_model_pipe = Pipeline(steps=[
    ('ct', ct),
    ('xgb', XGBClassifier(random_state = 138))
])

In [47]:
xgb_model_pipe.fit(X_train, y_train)

xgb_model_pipe.score(X_train, y_train)

0.8331323182069451

### Score comparison

- `dummy` - 0.54209
- `lr` (best) - 0.72717
- `rfc` (best) - 0.84283
- `etc` (best) - 0.90123
- `gbc` - 0.75375
- `xgb` - 0.83504

## Final Model - stacking `RandomForest` and `XGBoost`

In [48]:
final_estimators = [
    ('rfc_model', rfc_pipe_three),
    ('xgb_model', xgb_model_pipe)
]

final_model_pipe = StackingClassifier(final_estimators)

In [49]:
final_model_pipe.fit(X_train, y_train)

StackingClassifier(estimators=[('rfc_model',
                                Pipeline(steps=[('ct',
                                                 ColumnTransformer(remainder='passthrough',
                                                                   transformers=[('subpipe_num',
                                                                                  Pipeline(steps=[('num_impute',
                                                                                                   SimpleImputer(strategy='median')),
                                                                                                  ('ss',
                                                                                                   StandardScaler())]),
                                                                                  ['amount_tsh',
                                                                                   'gps_height',
                                             

In [50]:
final_model_pipe.score(X_train, y_train)

0.8507965224383135

In [51]:
precision_score(y_true = y_train,
                y_pred = final_model_pipe.predict(X_train),
                average = 'weighted')

0.855267568440182

In [52]:
recall_score(y_true = y_train,
             y_pred = final_model_pipe.predict(X_train),
             average = 'weighted')

0.8507965224383135

In [53]:
cross_val_score(estimator = final_model_pipe,
                X = X_train,
                y = y_train)

array([0.79020101, 0.78592965, 0.78756281, 0.78464631, 0.79469783])

Looks like our model is overfit a bit.

In [54]:
print(classification_report(y_true = y_train,
                            y_pred = final_model_pipe.predict(X_train)))

                         precision    recall  f1-score   support

             functional       0.82      0.95      0.88     21574
functional needs repair       0.85      0.36      0.51      2921
         non functional       0.90      0.81      0.85     15303

               accuracy                           0.85     39798
              macro avg       0.86      0.71      0.75     39798
           weighted avg       0.86      0.85      0.84     39798



**Final `f1-score` per status group** (on training data):
- `functional` - 0.88
- `functional needs repair` - 0.51
- `non functional` - 0.85

***Accuracy score:*** 0.85

### Unseen data (`test` sets)

In [55]:
final_model_pipe.score(X_test, y_test)

0.7909907152331395

In [56]:
precision_score(y_true = y_test,
                y_pred = final_model_pipe.predict(X_test),
                average = 'weighted')

0.787371446814695

In [57]:
recall_score(y_true = y_test,
             y_pred = final_model_pipe.predict(X_test),
             average = 'weighted')

0.7909907152331395

In [58]:
print(classification_report(y_true = y_test,
                            y_pred = final_model_pipe.predict(X_test)))

                         precision    recall  f1-score   support

             functional       0.77      0.91      0.83     10685
functional needs repair       0.62      0.22      0.33      1396
         non functional       0.84      0.73      0.78      7521

               accuracy                           0.79     19602
              macro avg       0.74      0.62      0.65     19602
           weighted avg       0.79      0.79      0.78     19602



**Final `f1-score` per status group** (on test data):
- `functional` - 0.83
- `functional needs repair` - 0.33
- `non functional` - 0.78

***Precision score* for `non functional` predictions**: 0.84

***Accuracy score:*** 0.79